pilot/pkg/xds/eds.go

```diff
@@ -462,11 +474,19 @@ func (eds *EdsGenerator) buildEndpoints(proxy *model.Proxy,
 			}
 		}
 		builder := NewEndpointBuilder(clusterName, proxy, req.Push)
-		if marshalledEndpoint, f := eds.Server.Cache.Get(builder); f && !features.EnableUnsafeAssertions {
-			// We skip cache if assertions are enabled, so that the cache will assert our eviction logic is correct
-			resources = append(resources, marshalledEndpoint)
-			cached++
-		} else {
+
+		marshalledEndpoint, token := eds.Server.Cache.Get(&builder)
+		if marshalledEndpoint != nil && !features.EnableUnsafeAssertions {
+			// If PA does not change or it has changed but the eds cache has already been refreshed in this period.
+			if useCache || (!useCache && token >= model.CacheToken(req.Start.UnixNano())) {
+				// We skip cache if assertions are enabled, so that the cache will assert our eviction logic is correct
```

Comment:
@dddddai @howardjohn This is to address the concern of not using the eds built in the same pushing period.